### ISSUE TITLE
Constructing multi-k Bloom filter de Bruijn graphs?

### ISSUE THREAD

**Issue by dbrowneup** [bcgsc/abyss#102]
Hi guys,

I found this paper recently, which I suspect you are probably already aware of, since it's not exactly a new paper (published 2012): [Scaling metagenome sequence assembly with probabilistic de Bruijn graphs](http://www.pnas.org/content/109/33/13272.abstract).

There's a particular quote in this paper that captured my attention (in the concluding remarks)...

"Third, the memory usage is independent of the k-mer size chosen, making this representation convenient for exploring properties at many different parameters. **It also allows the storage and traversal of de Bruijn graphs at multiple k-mer sizes within a single structure**, although we have not yet explored these properties."

I'm just wondering if this is something that might be possible with `abyss-bloom`? I've begun to toy around with `abyss-sealer` to fill gaps, with limited success so far. But it was apparent to me that the Bloom filters are constructed serially for each k-mer size. Yet the paper above seems to indicate you could construct a Bloom filter with multiple k-mers in parallel inside a single data structure.

**Comment by sjackman** [comment]
Yes, I am definitely familiar with that paper. It's technically possible, and, at least in principle, not terribly difficult, at least at the level of the data structure. To really take advantage of it though requires designing algorithms that adaptively search multiple values of *k*, by moving up to larger values of *k* in repetitive ares, and moving down to lower values of *k* in areas of low coverage, dynamically. We haven't done this work. I believe there's at least one paper on this topic, though I can't seem to find it now.

**Comment by dbrowneup** [comment]
Is this perhaps the paper you were thinking of? Closest thing I could find to what we are talking about: http://ieeexplore.ieee.org/xpl/articleDetails.jsp?arnumber=7149295

Here's the corresponding code that is described in the paper: https://github.com/alexbowe/cosmo

**Comment by sjackman** [comment]
I believe it was an older paper, but thanks for pointing me to @alexbowe's paper. It was a good read! Nice work, Alex. I hope that I can find some time to experiment with Cosmo.
Alex, do you recall a paper that discussed algorithms for traversing a multiorder dBG? It might have been a few years back, and I believe it was more focused on algorithms than data structures.

**Comment by alexbowe** [comment]
Thanks @sjackman :) I am slowly working on a BGL-compatible interface for Cosmo as per your recommendation (I've been meaning to talk to you about it actually...).

As for the paper, I don't know it sorry. If you find it I'd love to read it :) We have kicked around a few ideas, but nothing concrete. It'd be nice to put this variable order stuff to use...

**Comment by sjackman** [comment]
A BGL interface would be great. It would be fun to try hooking the Cosmo data structure into ABySS. Feel free to fire me an e-mail, or we can chat over another medium some time.

I didn't have any luck finding the paper on Google Scholar. It may be that I was thinking of a conference poster.

**Comment by dbrowneup** [comment]
Wish I could help code this up, but I'm afraid all I know is a little Python! However, if you guys put something together in the near future, I'd be glad to help test it! :)

**Comment by alexbowe** [comment]
@sjackman, you might be interested in this: @mikenz1976 wrote a BGL-adapter for the fixed-order dBG: https://github.com/mikenz1976/cosmo/blob/master/bgl_sdb_adapter.hpp (example here: https://github.com/mikenz1976/cosmo/blob/master/bgl_sdb_adapter_test.cpp).

After I finish the current things I'm working on, I will merge it and try to write a variable-order one. Probably something like `max_order(G)`, `get_order(node)`, and `set_order(G, node, new_order)` (and edge iterators will work as expected after an order change).

**Comment by mikenz1976** [comment]
Hi @sjackman...
Is it a case of the variable order de Bruijn needing to support a certain set of BGL [concepts](http://www.boost.org/doc/libs/1_60_0/libs/graph/doc/graph_concepts.html) & [properties](http://www.boost.org/doc/libs/1_60_0/libs/graph/doc/using_adjacency_list.html#sec:adjacency-list-properties) in order for it to be easily integrated into abyss?
Mike

**Comment by sjackman** [comment]
> Is it a case of the variable order de Bruijn needing to support a certain set of BGL concepts & properties in order for it to be easily integrated into abyss?

For ABySS to make use of the fixed-order dBG in Cosmo is reasonably straight forward. There's probably some worked involved to make the two play nicely together, for example, in mapping vertices to unique indices and their properties, but it should all be very doable.

A lot of this work was already done as a proof of concept, although unfortunately never merged into the trunk, for the paper [*On the representation of de Bruijn graphs*](http://doi.org/10.1007/978-3-319-05269-4_4) in the branch [dbgfm](https://github.com/bcgsc/abyss/tree/dbgfm/dbgfm) of ABySS.

ABySS has no algorithms to make use of the variable order dBG yet.

**Comment by dbrowneup** [comment]
Hi guys,

I came across some neat code that is relevant to this topic: https://github.com/JohnLonginotto/ACGTrie

There's an explanation of the data structure here: https://github.com/JohnLonginotto/ACGTrie/blob/master/docs/OUTPUT.md

I think this "warp table" is a really nice data structure, though one would still need to write traversal algorithms capable of utilizing the information it contains.

**Comment by dbrowneup** [comment]
Quick follow up... Googling for "Bloom filter trie" led me to a recently published paper that y'all might find interesting: https://almob.biomedcentral.com/articles/10.1186/s13015-016-0066-8

Code: https://github.com/GuillaumeHolley/BloomFilterTrie

**Comment by sjackman** [comment]
ACGTrie is a cool idea. It's a normal trie but stored in a table rather than a tree. I'd be curious to know how the memory/speed compares to an FM-index, which supports the same operations.

**Comment by sjackman** [comment]
Bloom Filter Trie looks interesting. That one's going to take some time though to grok.
